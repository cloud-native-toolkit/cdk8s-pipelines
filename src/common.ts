const paramNameRegex = new RegExp('^[a-zA-Z_][-a-zA-Z0-9_]+$');

function hasValidName(res: NamedResource): boolean {
  let valid = false;
  if (res.name) {
    valid = paramNameRegex.test(res.name);
  }
  return valid;
}

export interface NamedResource {
  readonly name?: string;
}
