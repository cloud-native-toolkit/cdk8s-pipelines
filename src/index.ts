export * from './pipelines';
export * from './tasks';
export * from './common';
