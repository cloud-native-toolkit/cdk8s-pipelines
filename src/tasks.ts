/**
 * This file includes the basic objects for creating Tasks and dependencies.
 * While cdk8s supports generating this file from the CRD, the Tekton CRD at
 * the moment does not have the fields specified. See https://storage.googleapis.com/tekton-releases/pipeline/latest/release.yaml,
 * specfically here:
 *       schema:
 *         openAPIV3Schema:
 *           type: object
 *           # OpenAPIV3 schema allows Kubernetes to perform validation on the schema fields
 *           # and use the schema in tooling such as `kubectl explain`.
 *           # Using "x-kubernetes-preserve-unknown-fields: true"
 *           # at the root of the schema (or within it) allows arbitrary fields.
 *           # We currently perform our own validation separately.
 *           # See https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema
 *           # for more info.
 *           x-kubernetes-preserve-unknown-fields: true
 * For the time being, these objects are created manually using the latest
 * Tekton documentation as a reference.
 */
import { ApiObject, ApiObjectMetadata, GroupVersionKind } from 'cdk8s';
import { Construct } from 'constructs';
import { NamedResource } from './common';

/**
 * A workspace used by a Task. See https://tekton.dev/docs/pipelines/workspaces/#using-workspaces-in-tasks for more information.
 */
export interface TaskWorkspace extends NamedResource {
  readonly workspace?: string;
}

export interface TaskWorkspaceRef extends TaskWorkspace {
  readonly description?: string;
}

export class TaskRef {
  name?: string;

  constructor(name: string) {
    this.name = name;
  }
}

/**
 * A Task parameter value.
 */
export interface TaskParam extends NamedResource {
  /**
   * The value of the task parameter.
   */
  readonly value?: string;
}

/**
 * Specifies execution parameters for the Task.
 */
export interface TaskSpecParam extends NamedResource {
  readonly type?: string;
  readonly description?: string;
  readonly default?: string;
}

/**
 * The volume mount for the task.
 */
export interface TaskVolumeMount extends NamedResource {
  readonly mountPath?: string;
}

/**
 * The step for a Task. See https://tekton.dev/docs/pipelines/tasks/#defining-steps
 * @see https://tekton.dev/docs/pipelines/container-contract/
 */
export interface TaskStep extends NamedResource {
  /**
   * The name of the container image to use for the Step.
   */
  readonly image?: string;
  /**
   * A script that will be executed on the image. If `script` is specified,
   * then the `command` value should not be specified.
   */
  readonly script?: string;
  /**
   * The command and its arguments (provided in the form of an array) to
   * execute on the container. If `command` is supplied, you should not supply
   * `script`.
   */
  readonly command?: string[];
  /**
   * Alternatively, you can supply `args` to the `command` value here.
   */
  readonly args?: string[];
  /**
   * The volume mounts to use for the task.
   */
  readonly volumeMounts?: TaskVolumeMount[];
  /**
   * The name of the working directory for the step
   */
  readonly workingDir?: string;
}

/**
 * The Task spec.
 *
 * @see https://tekton.dev/docs/pipelines/tasks/#configuring-a-task
 */
export interface TaskSpec {
  /**
   * The description of the `Task`.
   * @see https://tekton.dev/docs/pipelines/tasks/#adding-a-description
   */
  readonly description?: string;
  /**
   * The `Task`'s parameters.
   * @see https://tekton.dev/docs/pipelines/tasks/#specifying-parameters
   */
  readonly params?: TaskSpecParam[];
  /**
   * The steps that will be executed as part of the Task. The `Step` should
   * fit the (container contract)[https://tekton.dev/docs/pipelines/container-contract/]
   * @see https://tekton.dev/docs/pipelines/tasks/#defining-steps
   */
  readonly steps?: TaskStep[];
}

/**
 * Properties used to create the Task.
 */
export interface TaskProps {
  /**
   * The object [metadata](https://kubernetes.io/docs/concepts/overview/working-with-objects/#required-fields)
   * that conforms to standard Kubernetes metadata.
   */
  readonly metadata?: ApiObjectMetadata;
  /**
   * The `spec` is the configuration of the `Task` object.
   */
  readonly spec?: TaskSpec;
}

/**
 * A Tekton Task, which is
 * > a collection of Steps that you define and arrange in
 * > a specific order of execution as part of your continuous integration flow. A
 * > Task executes as a Pod on your Kubernetes cluster. A Task is available within a
 * > specific namespace, while a ClusterTask is available across the entire
 * > cluster.
 *
 * @see https://tekton.dev/docs/pipelines/tasks/
 */
export class Task extends ApiObject {
  /**
   * Returns the apiVersion and kind for "Task"
   */
  public static readonly GVK: GroupVersionKind = {
    apiVersion: 'tekton.dev/v1beta1',
    kind: 'Task',
  };

  /**
   * Renders a Kubernetes manifest for "Task".
   *
   * This can be used to inline resource manifests inside other objects (e.g. as templates).
   *
   * @param props initialization props
   */
  public static manifest(props: TaskProps = {}): any {
    return {
      ...Task.GVK,
      ...props,
    };
  }

  private readonly _metadata?: ApiObjectMetadata;
  private readonly _spec?: TaskSpec;

  /**
   * Defines a "Task" API object
   * @param scope the scope in which to define this object
   * @param id a scope-local name for the object
   * @param props initialization props
   */
  public constructor(scope: Construct, id: string, props: TaskProps = {}) {
    super(scope, id, {
      ...Task.GVK,
      ...props,
    });
    this._metadata = props.metadata;
    this._spec = props.spec;
  }

  /**
   * Renders the object to Kubernetes JSON.
   */
  public toJson(): any {
    const result = {
      ...Task.GVK,
      ...{
        metadata: this._metadata,
        spec: this._spec,
      },
    };
    return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({
      ...r,
      [i[0]]: i[1],
    }), {});
  }
}
