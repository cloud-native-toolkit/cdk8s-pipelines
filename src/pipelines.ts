import { ApiObject, ApiObjectMetadata, GroupVersionKind } from 'cdk8s';
import { Construct } from 'constructs';
import { NamedResource } from './common';
import { TaskParam, TaskRef, TaskWorkspace, TaskWorkspaceRef } from './tasks';

// The following interfaces and classes are strictly for generating the YAML or
// JSON in the proper format for the Tekton pipelines. See the builders to use
// classes and a fluid-style syntax that provides easier building and checking.

/**
 * A Pipeline parameter. See https://tekton.dev/docs/pipelines/pipelines/#specifying-parameters
 */
export interface PipelineParam extends NamedResource {
  readonly type?: string;
  readonly default?: string;
}

/**
 * A task in a pipeline. See https://tekton.dev/docs/pipelines/pipelines/#adding-tasks-to-the-pipeline
 */
export interface PipelineTask extends NamedResource {
  readonly taskRef?: TaskRef;
  readonly params?: TaskParam[];
  readonly runAfter?: string[];
  readonly workspaces?: TaskWorkspace[];
}

/**
 * A workspace for a pipeline. See https://tekton.dev/docs/pipelines/pipelines/#specifying-workspaces
 * and https://tekton.dev/docs/pipelines/workspaces/#using-workspaces-in-pipelines.
 */
export interface PipelineWorkspace extends NamedResource {
  /**
   * The description of the workspace.
   */
  readonly description?: string;
}

/**
 * A task definition at the pipeline level.
 */
export interface PipelineTaskDef extends PipelineTask {
  readonly refParams?: PipelineParam[];
  readonly refWorkspaces?: TaskWorkspaceRef[];
}

/**
 * The `spec` part of the Pipeline.
 *
 * @see https://tekton.dev/docs/pipelines/pipelines/
 */
export interface PipelineSpec {
  readonly description?: string;
  readonly params?: PipelineParam[];
  /**
   * The `tasks` are required on the Pipeline.
   */
  readonly tasks: PipelineTask[];
  readonly workspaces?: PipelineWorkspace[];
}

/**
 * Properties used to create the Pipelines.
 */
export interface PipelineProps {
  readonly metadata?: ApiObjectMetadata;
  readonly spec?: PipelineSpec;
}

/**
 *
 *
 * @schema Pipeline
 */
export class Pipeline extends ApiObject {

  /**
   * Returns the apiVersion and kind for "Pipeline"
   */
  public static readonly GVK: GroupVersionKind = {
    apiVersion: 'tekton.dev/v1beta1',
    kind: 'Pipeline',
  };

  private readonly _metadata?: ApiObjectMetadata;
  private readonly _spec?: PipelineSpec;

  /**
   * Renders a Kubernetes manifest for "Pipeline".
   *
   * This can be used to inline resource manifests inside other objects (e.g. as templates).
   *
   * @param props initialization props
   */
  // public static manifest(props: PipelineProps = {}): any {
  //   return {
  //     ...Pipeline.GVK,
  //     ...
  //   };
  // }

  /**
   * Defines a "Pipeline" API object
   * @param scope the scope in which to define this object
   * @param id a scope-local name for the object
   * @param props initialization props
   */
  public constructor(scope: Construct, id: string, props: PipelineProps = {}) {
    super(scope, id, {
      ...Pipeline.GVK,
    });
    this._metadata = props.metadata;
    this._spec = props.spec;
  }


  /**
   * Renders the object to Kubernetes JSON.
   */
  public toJson(): any {
    const result = {
      ...Pipeline.GVK,
      ...{
        metadata: this._metadata,
        spec: this._spec,
      },
    };
    return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({
      ...r,
      [i[0]]: i[1],
    }), {});
  }
}
