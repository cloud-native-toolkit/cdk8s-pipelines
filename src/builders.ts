/**
 * This file has builders in it for the various pipeline constructs.
 */


import * as fs from 'fs';
import { ApiObject } from 'cdk8s';
import { Construct } from 'constructs';
import { Pipeline, PipelineParam, PipelineTask, PipelineTaskDef, PipelineTaskWorkspace, PipelineWorkspace } from './pipelines';
import { Task, TaskEnvValueSource, TaskParam, TaskProps, TaskRef, TaskSpecParam, TaskStep, TaskStepEnv, TaskWorkspace } from './tasks';

/**
 * The workspace record (WorkspaceRec) is used by the PipelineTaskBuilder
 * internally to hold information about the parameters.
 */
export interface WorkspaceRec {
  readonly name?: string;
  readonly refName?: string;
  readonly description?: string;
}

/**
 * The parameter record (ParamRec) is used by the PipelineTaskBuilder builder
 * internally to hold values for the parameter reference.
 */
export interface ParamRec {
  readonly name?: string;
  readonly description?: string;
  readonly defaultValue?: string;
  readonly refName?: string;
  readonly type?: string;
  readonly value?: string;
}


export class TaskStepBuilder {
  private _url?: string;
  private _obj?: ApiObject;
  private _name?: string;
  private _dir?: string;
  private _image?: string;
  private _cmd?: string[];
  private _args?: string[];
  private _env?: TaskStepEnv[];

  /**
   *
   */
  public constructor() {

  }

  /**
   * The name of the `Step` of the `Task`.
   */
  public get name(): string | undefined {
    return this._name;
  }

  /**
   * The name of the container `image` used to execute the `Step` of the
   * `Task`.
   */
  public get image(): string | undefined {
    return this._image;
  }

  /**
   * Gets the URL from which the script data should be loaded, if it is defined.
   */
  public get scriptUrl(): string | undefined {
    return this._url;
  }

  /**
   * Gets the object that is used for the `script` value, if there is one
   * defined.
   */
  public get scriptObj(): ApiObject | undefined {
    return this._obj;
  }

  /**
   * Gets the command-line arguments that will be supplied to the `command`.
   */
  public get args(): string[] | undefined {
    return this._args;
  }

  /**
   * Gets the command used for the `Step` on the `Task`.
   */
  public get command(): string[] | undefined {
    return this._cmd;
  }

  public get workingDir(): string | undefined {
    return this._dir;
  }

  public withName(name: string): TaskStepBuilder {
    this._name = name;
    return this;
  }

  /**
   * The name of the image to use when executing the `Step` on the `Task`
   * @param img
   */
  public withImage(img: string): TaskStepBuilder {
    this._image = img;
    return this;
  }

  /**
   * The name of the command to use when running the `Step` of the `Task`. If
   * `command` is specified, do not specify `script`.
   * @param cmd
   */
  public withCommand(cmd: string[]): TaskStepBuilder {
    this._cmd = cmd;
    return this;
  }

  /**
   * The args to use with the `command`.
   * @param args
   */
  public withArgs(args: string[]): TaskStepBuilder {
    this._args = args;
    return this;
  }

  /**
   * If supplied, uses the content found at the given URL for the
   * `script` value of the step. Use this as an alternative to "heredoc", which
   * is embedding hard-coded shell or other scripts in the step.
   *
   * If you supply this, do not supply a value for `fromScriptObject`.
   * @param url
   */
  public fromScriptUrl(url: string): TaskStepBuilder {
    this._url = url;
    return this;
  }

  /**
   * If supplied, uses the cdk8s `ApiObject` supplied as the body of the
   * `script` for the `Task`. This is most useful when used with `oc apply` or
   * other tasks in which you want to apply the object during the step in the
   * pipeline.
   *
   * If you supply this, do not supply a value for `fromScriptUrl`.
   * @param obj
   */
  public fromScriptObject(obj: ApiObject): TaskStepBuilder {
    this._obj = obj;
    return this;
  }

  /**
   * The `workingDir` of the `Task`.
   * @param dir
   */
  public withWorkingDir(dir: string): TaskStepBuilder {
    this._dir = dir;
    return this;
  }

  withEnv(name: string, valueFrom: TaskEnvValueSource): TaskStepBuilder {
    if (!this._env) {
      this._env = new Array<TaskStepEnv>();
    }
    this._env.push({
      name: name,
      valueFrom: valueFrom,
    });
    return this;
  }

  public buildTaskStep(): TaskStep | undefined {
    if (this.scriptUrl) {
      if (this.scriptObj) {
        throw new Error('Cannot specify both a URL source and an object source for the script.');
      }
      // Load the script from the URL location and use it
      const data = fs.readFileSync(this.scriptUrl, {
        encoding: 'utf8',
        flag: 'r',
      });

      const lines = data.replace(/\n/g, '\\n');
      if (data) {
        return {
          name: this.name,
          image: this.image,
          script: lines,
          workingDir: this.workingDir,
          env: this._env,
        };
      }
    } else {
      return {
        name: this.name,
        image: this.image,
        command: this.command,
        args: this.args,
        workingDir: this.workingDir,
        env: this._env,
      };
    }
    return undefined;
  }


}

/**
 * This is the builder for creating Tekton `Task` objects that are independent
 * of a `Pipeline`. They
 *
 * To use a builder for tasks that will be used in a Pipeline, use the
 * `PipelineBuilder` instead.
 */
export class TaskBuilder {

  private readonly _scope?: Construct;
  private readonly _id?: string;
  private _steps?: TaskStepBuilder[];
  private _name?: string;
  private _description?: string;
  private _workspaces?: TaskWorkspace[];
  private _params?: TaskSpecParam[];

  /**
   * Creates a new instance of the `TaskBuilder` using the given `scope` and
   * `id`.
   * @param scope
   * @param id
   */
  public constructor(scope: Construct, id: string) {
    this._scope = scope;
    this._id = id;
    // These are required, and it's better to just create it rather than
    // check each time.
    this._steps = new Array<TaskStepBuilder>();
  }

  /**
   * Gets the name of the `Task` built by the `TaskBuilder`.
   */
  public get name(): string | undefined {
    return this._name;
  }

  /**
   * Sets the name of the `Task` being built.
   * @param name
   */
  public withName(name: string): TaskBuilder {
    this._name = name;
    return this;
  }

  /**
   * Sets the `description` of the `Task` being built.
   * @param description
   */
  public withDescription(description: string): TaskBuilder {
    this._description = description;
    return this;
  }

  /**
   * Gets the `description` of the `Task`.
   */
  public get description(): string | undefined {
    return this._description;
  }

  /**
   * Adds the specified workspace to the `Task`.
   * @param name
   * @param description
   */
  public withWorkspace(name: string, description: string = ''): TaskBuilder {
    if (! this._workspaces) {
      this._workspaces = new Array<TaskWorkspace>();
    }
    this._workspaces.push({
      name: name,
      description: description,
    });
    return this;
  }

  public withStringParam(name: string, description: string = '', defaultValue: string = '') : TaskBuilder {
    if (! this._params) {
      this._params = new Array<TaskParam>();
    }
    this._params.push({
      name: name,
      description: description,
      default: defaultValue,
    });
    return this;
  }

  /**
   * Adds the given `step` (`TaskStepBuilder`) to the `Task`.
   * @param step
   */
  public withStep(step: TaskStepBuilder): TaskBuilder {
    this._steps!.push(step);
    return this;
  }

  /**
   * Builds the `Task`.
   */
  public buildTask(): void {

    const taskSteps = new Array<TaskStep>();

    this._steps?.forEach((s) => {
      const step = s.buildTaskStep();
      if (step) {
        taskSteps.push(step);
      }
    });

    const props: TaskProps = {
      metadata: {
        name: this.name,
      },
      spec: {
        description: this.description,
        workspaces: this._workspaces,
        params: this._params,
        steps: taskSteps,
      },
    };

    new Task(this._scope!, this._id!, props);

  }
}

/**
 * The PipelineTaskBuilder creates a PipelineTaskDef, which is used to
 * intelligently add a task and its properties to a Pipeline.
 */
export class PipelineTaskBuilder {

  /**
   * Creates an instance of this builder with the provided `TaskBuilder` as
   * a source.
   * @param task
   */
  public static createFromTask(task: TaskBuilder) {
    const created = new PipelineTaskBuilder();
    created.withName(task.name!);

    return created;
  }

  private _name?: string;
  private _taskRefName?: string;
  private _workspaces?: WorkspaceRec[];
  private _parameters?: ParamRec[];
  private _after?: string[];

  /**
   * Creates a new instance of the PipelineTaskBuilder
   */
  public constructor() {
  }

  /**
   * Provides the name for the pipeline task and will be
   * rendered as the `name` property.
   * @param name
   */
  public withName(name: string): PipelineTaskBuilder {
    this._name = name;
    return this;
  }

  /**
   * Returns the name of the task being built by the task builder.
   */
  public get name(): string {
    return this._name!;
  }

  /**
   * Returns the parameter records for the task builder.
   */
  public get parameters(): ParamRec[] {
    return this._parameters!;
  }

  /**
   * Returns the workspace records for the task builder.
   */
  public get workspaces(): WorkspaceRec[] {
    return this._workspaces!;
  }

  /**
   * Returns the task reference name for the task builder.
   */
  public get taskReference(): string {
    return this._taskRefName!;
  }

  /**
   * Allows you to specify that this task should be completed after another
   * task. Can be called multiple times to add multiple tasks.
   * @param otherTaskB
   */
  public doAfter(otherTaskB: PipelineTaskBuilder): PipelineTaskBuilder {
    if (!this._after) {
      this._after = new Array<string>();
    }
    this._after.push(otherTaskB.name);
    return this;
  }

  /**
   * Creates and adds a [task reference](https://tekton.dev/docs/pipelines/tasks/#configuring-a-task) to the
   * task, using the value supplied as the name of the reference. It returns
   * a reference to the builder.
   * @param ref The name of the task that is referenced.
   */
  public withTaskReference(ref: string): PipelineTaskBuilder {
    this._taskRefName = ref;
    return this;
  }


  /**
   * Adds workspace information to the pipeline task and returns a
   * reference to the builder.
   * @param name The name of the workspace on the pipeline task.
   * @param refName The `workspace` value, which will be the name of the workspace
   *  at the Pipeline level.
   * @param description The description of the Pipeline workspace.
   */
  public withWorkspace(name: string, refName: string, description: string): PipelineTaskBuilder {
    if (!this._workspaces) {
      this._workspaces = new Array<WorkspaceRec>();
    }
    this._workspaces.push({
      name: name,
      refName: refName,
      description: description,
    });
    return this;
  }

  /**
   * Adds a parameter of type string to both the task and the pipeline itself.
   * @param name The name of the param on the task.
   * @param refName The name of the param at the pipeline level.
   * @param value The value of the parameter on the task.
   * @param description The default value for the param at the pipeline level
   * @param defaultValue The default value for the param at the pipeline level
   */
  public withStringParam(name: string, refName: string, value: string, description: string = '', defaultValue: string = ''): PipelineTaskBuilder {
    if (!this._parameters) {
      this._parameters = new Array<ParamRec>();
    }
    this._parameters!.push({
      name: name,
      value: value,
      refName: refName,
      type: 'string',
      defaultValue: defaultValue,
      description: description,
    });
    return this;
  }

  /**
   * Builds the Pipeline
   */

  public buildPipelineTask(): PipelineTaskDef {
    // TODO: An important part of build here is to assert that the
    // object has been correctly made and throw very meaningful errors.
    const pt = {
      name: this.name,
      taskRef: new TaskRef(this._taskRefName!),
      params: [
        {
          name: 'url',
          value: '$(params.repo-url)',
        },
      ],
      refParams: [
        {
          name: 'repo-url',
          default: '',
          type: 'string',
        },
      ],
      refWorkspaces: [
        {
          name: 'output',
          workspace: 'shared-files',
          description: 'The files that are cloned by the task',
        },
      ],
    };
    return pt;
  }

}

export class PipelineBuilder {
  private readonly _scope?: Construct;
  private readonly _id?: string;
  private _name?: string;
  private _description?: string;
  private _tasks?: PipelineTaskBuilder[];

  public constructor(scope: Construct, id: string) {
    this._scope = scope;
    this._id = id;
  }

  /**
   * Provides the name for the pipeline task and will be
   * rendered as the `name` property.
   * @param name
   */
  public withName(name: string): PipelineBuilder {
    this._name = name;
    return this;
  }

  /**
   * Gets the name of the pipeline
   */
  public get name(): string {
    return this._name!;
  }

  /**
   * Provides the name for the pipeline task and will be
   * rendered as the `name` property.
   * @param description
   */
  public withDescription(description: string): PipelineBuilder {
    this._description = description;
    return this;
  }

  // Adds the task to the pipeline.
  public withTask(taskB: PipelineTaskBuilder): PipelineBuilder {
    // Add the task to the list of tasks...
    if (!this._tasks) {
      this._tasks = new Array<PipelineTaskBuilder>();
    }
    this._tasks.push(taskB);
    return this;
  }

  /**
   * Builds the actual [Pipeline]() from the settings configured using the
   * fluid syntax.
   */
  public buildPipeline(): void {
    // TODO: validate the object
    const pipelineParams: PipelineParam[] = new Array<PipelineParam>();
    const pipelineWorkspaces: PipelineWorkspace[] = new Array<PipelineWorkspace>();
    const pipelineTasks: PipelineTask[] = new Array<PipelineTask>();

    this._tasks?.forEach(t => {

      const taskParams: TaskParam[] = new Array<TaskParam>();
      const taskWorkspaces: PipelineTaskWorkspace[] = new Array<TaskWorkspace>();

      t.parameters?.forEach(p => {
        const pp = pipelineParams.find((value, index, obj) => value.name == obj[index].name);
        if (!pp) {
          pipelineParams.push({
            name: p.refName,
            type: p.type,
          });
        }

        taskParams.push({
          name: p.name,
          value: p.value,
        });
      });

      t.workspaces.forEach((w) => {
        // Only add the workspace on the pipeline level if it is not already
        // there...
        const ws = pipelineWorkspaces.find((value, index, obj) => value.name == obj[index].name);
        if (!ws) {
          pipelineWorkspaces.push({
            name: w.refName,
            description: w.description,
          });
        }

        taskWorkspaces.push({
          name: w.name,
          workspace: w.refName,
        });
      });

      pipelineTasks.push({
        name: t.name,
        taskRef: {
          name: t.taskReference,
        },
        params: taskParams,
        workspaces: taskWorkspaces,
      });
    });

    new Pipeline(this._scope!, this._id!, {
      metadata:
        {
          name: this.name,
        },
      spec: {
        description: this._description,
        params: pipelineParams,
        workspaces: pipelineWorkspaces,
        tasks: pipelineTasks,
      },
    });
  }
}
