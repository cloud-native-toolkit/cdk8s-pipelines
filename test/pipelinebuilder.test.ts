import { Chart, Testing } from 'cdk8s';
import { ChartProps } from 'cdk8s/lib/chart';
import { Construct } from 'constructs';
// @ts-ignore
import { Pipeline, PipelineBuilder, PipelineTaskBuilder, PipelineTaskDef, TaskRef, TaskBuilder, WorkspaceBuilder, ParameterBuilder } from '../src';

class MyTestChart extends Chart {
  constructor(scope: Construct, id: string, props?: ChartProps) {
    super(scope, id, props);

    const myTask = new TaskBuilder(this, 'git-clone')
      .withName('fetch-source')
      .withWorkspace(new WorkspaceBuilder('output')
        .withName('shared-data')
        .withDescription('The files cloned by the task'))
      .withStringParam(new ParameterBuilder('url').withPiplineParameter('repo-url', ''));

    new PipelineBuilder(this, 'my-pipeline')
      .withName('clone-build-push')
      .withDescription('This pipeline closes a repository, builds a Docker image, etc.')
      .withTask(myTask)
      .buildPipeline();
  }
}

class MySecondTestChart extends Chart {
  constructor(scope: Construct, id: string, props?: ChartProps) {
    super(scope, id, props);

    const myTask = new TaskBuilder(this, 'git-clone')
      .withName('fetch-source')
      .withWorkspace(new WorkspaceBuilder('output')
        .withName('shared-data')
        .withDescription('The files cloned by the task'))
      .withWorkspace(new WorkspaceBuilder('ssh-credentials')
        .withName('ssh-creds')
        .withDescription('The SSH files for credentials'))
      .withWorkspace(new WorkspaceBuilder('config')
        .withName('config-data')
        .withDescription('The files for configuration for stuff'))
      .withStringParam(new ParameterBuilder('url').withPiplineParameter('repo-url', ''))
      .withStringParam(new ParameterBuilder('name').withPiplineParameter('your-name', ''))
      .withStringParam(new ParameterBuilder('color').withPiplineParameter('your-color', ''))
      .withStringParam(new ParameterBuilder('quest').withPiplineParameter('your-quest', ''));

    new PipelineBuilder(this, 'my-test-pipeline-2')
      .withName('clone-build-push')
      .withDescription('This pipeline closes a repository, builds a Docker image, etc.')
      .withTask(myTask)
      .buildPipeline();
  }
}

class MyTestChartWithDups extends Chart {
  constructor(scope: Construct, id: string, props?: ChartProps) {
    super(scope, id, props);

    const myWorkspace = new WorkspaceBuilder('output')
      .withDescription('The files cloned by the task')
      .withName('shared-data');

    const urlParam = new ParameterBuilder('url')
      .withPiplineParameter('repo-url', '');

    const myTask = new TaskBuilder(this, 'git-clone')
      .withName('fetch-source')
      .withWorkspace(myWorkspace)
      .withStringParam(urlParam)
    ;

    const myTask2 = new TaskBuilder(this, 'cat-readme')
      .withName('print-readme')
      .withWorkspace(myWorkspace)
      .withStringParam(urlParam)
    ;

    new PipelineBuilder(this, 'my-pipeline')
      .withName('clone-build-push')
      .withDescription('This pipeline closes a repository, builds a Docker image, etc.')
      .withTask(myTask)
      .withTask(myTask2)
      .buildPipeline();
  }
}

class MyTestChartWithStaticOverride extends Chart {

  constructor(scope: Construct, id: string, props?: ChartProps) {
    super(scope, id, props);

    const urlParam = new ParameterBuilder('url')
      .withPiplineParameter('repo-url', '');

    const myTask2 = new TaskBuilder(this, 'cat-readme')
      .withName('print-readme')
      .withStringParam(urlParam)
    ;

    // This should override the parameter with a static value.
    myTask2.withStringParam(new ParameterBuilder('url').withValue('https://api.example.io'));

    new PipelineBuilder(this, 'my-pipeline')
      .withName('clone-build-push')
      .withDescription('This pipeline closes a repository, builds a Docker image, etc.')
      .withTask(myTask2)
      .buildPipeline();
  }
}

describe('PipelineBuilderTest', () => {
  test('PipelineBuilder', () => {
    const app = Testing.app();
    const chart = new MyTestChart(app, 'test-chart');
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });

  test('PipelineBuilderWithComplexTasks', () => {
    const app = Testing.app();
    const chart = new MySecondTestChart(app, 'my-second-test-chart');
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });

  test('PipelineBuilderWithDuplicates', () => {
    const app = Testing.app();
    const chart = new MyTestChartWithDups(app, 'test-chart');
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });

  test('PipelineBuilderWithStaticOverride', () => {
    const app = Testing.app();
    const chart = new MyTestChartWithStaticOverride(app, 'test-chart');
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });
});
