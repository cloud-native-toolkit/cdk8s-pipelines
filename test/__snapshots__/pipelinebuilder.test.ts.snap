// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PipelineBuilderTest PipelineBuilder 1`] = `
[
  {
    "apiVersion": "tekton.dev/v1beta1",
    "kind": "Pipeline",
    "metadata": {
      "name": "clone-build-push",
    },
    "spec": {
      "description": "This pipeline closes a repository, builds a Docker image, etc.",
      "params": [
        {
          "name": "repo-url",
          "type": "string",
        },
      ],
      "tasks": [
        {
          "name": "fetch-source",
          "params": [
            {
              "name": "url",
              "value": "$(params.repo-url)",
            },
          ],
          "taskRef": {
            "name": "git-clone",
          },
          "workspaces": [
            {
              "name": "output",
              "workspace": "shared-data",
            },
          ],
        },
      ],
      "workspaces": [
        {
          "description": "The files cloned by the task",
          "name": "shared-data",
        },
      ],
    },
  },
]
`;

exports[`PipelineBuilderTest PipelineBuilderWithComplexTasks 1`] = `
[
  {
    "apiVersion": "tekton.dev/v1beta1",
    "kind": "Pipeline",
    "metadata": {
      "name": "clone-build-push",
    },
    "spec": {
      "description": "This pipeline closes a repository, builds a Docker image, etc.",
      "params": [
        {
          "name": "repo-url",
          "type": "string",
        },
        {
          "name": "your-name",
          "type": "string",
        },
        {
          "name": "your-color",
          "type": "string",
        },
        {
          "name": "your-quest",
          "type": "string",
        },
      ],
      "tasks": [
        {
          "name": "fetch-source",
          "params": [
            {
              "name": "url",
              "value": "$(params.repo-url)",
            },
            {
              "name": "name",
              "value": "$(params.your-name)",
            },
            {
              "name": "color",
              "value": "$(params.your-color)",
            },
            {
              "name": "quest",
              "value": "$(params.your-quest)",
            },
          ],
          "taskRef": {
            "name": "git-clone",
          },
          "workspaces": [
            {
              "name": "output",
              "workspace": "shared-data",
            },
            {
              "name": "ssh-credentials",
              "workspace": "ssh-creds",
            },
            {
              "name": "config",
              "workspace": "config-data",
            },
          ],
        },
      ],
      "workspaces": [
        {
          "description": "The files cloned by the task",
          "name": "shared-data",
        },
        {
          "description": "The SSH files for credentials",
          "name": "ssh-creds",
        },
        {
          "description": "The files for configuration for stuff",
          "name": "config-data",
        },
      ],
    },
  },
]
`;

exports[`PipelineBuilderTest PipelineBuilderWithDuplicates 1`] = `
[
  {
    "apiVersion": "tekton.dev/v1beta1",
    "kind": "Pipeline",
    "metadata": {
      "name": "clone-build-push",
    },
    "spec": {
      "description": "This pipeline closes a repository, builds a Docker image, etc.",
      "params": [
        {
          "name": "repo-url",
          "type": "string",
        },
      ],
      "tasks": [
        {
          "name": "fetch-source",
          "params": [
            {
              "name": "url",
              "value": "$(params.repo-url)",
            },
          ],
          "taskRef": {
            "name": "git-clone",
          },
          "workspaces": [
            {
              "name": "output",
              "workspace": "shared-data",
            },
          ],
        },
        {
          "name": "print-readme",
          "params": [
            {
              "name": "url",
              "value": "$(params.repo-url)",
            },
          ],
          "taskRef": {
            "name": "cat-readme",
          },
          "workspaces": [
            {
              "name": "output",
              "workspace": "shared-data",
            },
          ],
        },
      ],
      "workspaces": [
        {
          "description": "The files cloned by the task",
          "name": "shared-data",
        },
      ],
    },
  },
]
`;

exports[`PipelineBuilderTest PipelineBuilderWithStaticOverride 1`] = `
[
  {
    "apiVersion": "tekton.dev/v1beta1",
    "kind": "Pipeline",
    "metadata": {
      "name": "clone-build-push",
    },
    "spec": {
      "description": "This pipeline closes a repository, builds a Docker image, etc.",
      "params": [],
      "tasks": [
        {
          "name": "print-readme",
          "params": [
            {
              "name": "url",
              "value": "https://api.example.io",
            },
          ],
          "taskRef": {
            "name": "cat-readme",
          },
          "workspaces": [],
        },
      ],
      "workspaces": [],
    },
  },
]
`;
